### PR TITLE
Add DiffusionGemma block-diffusion SFT example

### PR DESCRIPTION
**Pull request by kashif**
Adds an SFT example for google/diffusiongemma-26B-A4B-it following the reference recipe: one response block per step, uniform random token corruption with t ~ U(eps, 1), two pass self conditioning, and flat cross entropy over the whole canvas plus an autoregressive co-loss on the encoder. LoRA targets the attention and dense MLP linears, matching the reference fine-tune.

Also adds a training chat template for DiffusionGemma (the stock template has no {% generation %} markers), registers it in get_training_chat_template, and adds trl-internal-testing/tiny-DiffusionGemmaForBlockDiffusion to the chat template CI along with its generation script.

Running the example needs transformers >= 5.12.0 (huggingface/transformers#46568, released in v5.12.0). The example trains as is. Gradient checkpointing is opt-in via --gradient_checkpointing and additionally needs huggingface/transformers#46572, which is still in review.

---

> [!NOTE]
> **Medium Risk**
> New training path and custom loss are isolated to an example script; core `SFTTrainer` behavior changes only via optional chat-template registration and test coverage for a new model type.
> 
> **Overview**
> Adds **block-diffusion SFT for DiffusionGemma** via a new `examples/scripts/sft_diffusion_gemma.py` that subclasses [`SFTTrainer`] with a custom `compute_loss`: random response-block canvas, uniform token corruption, optional two-pass self-conditioning, decoder cross-entropy plus encoder AR co-loss, and `--model_prediction_type` (`mean` vs `mean_loo`) for the LOO denoiser conversion. Defaults target GSM8K, reference LoRA/full-FT hyperparameters, and DeepSpeed ZeRO-3.
> 
> **Chat templates & assistant-only loss:** Bundles `diffusion_gemma.jinja` / `diffusion_gemma_training.jinja` (with `{% generation %}` markers) and wires them into `get_training_chat_template` so `assistant_only_loss=True` masks the assistant turn correctly.
> 
> **CI & tiny models:** Adds `for_block_diffusion/diffusion_gemma_for_block_diffusion.py` for `trl-internal-testing/tiny-DiffusionGemmaForBlockDiffusion`, extends chat-template tests to treat `ForBlockDiffusion` like VLMs, and documents the example in `example_overview.md`, `sft_trainer.md`, and `paper_index.md` (LOO paper).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8a0c9d6820a94bb95f7bb5463452cc58d28f4b08. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by kashif** [comment]
Updated now that the transformers side is merged: per example self conditioning, encoder AR co-loss, frozen router for full fine-tuning, and gradient checkpointing works with huggingface/transformers#46572.

File: examples/scripts/sft_diffusion_gemma.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+        input_ids = inputs["input_ids"]
+        labels = inputs["labels"]
+        device = input_ids.device
+        batch_size, seq_len = input_ids.shape
+        block_size = self.canvas_length
+
+        # The completion is the contiguous suffix of supervised tokens
+        supervised = labels != -100
+        prefix_len = supervised.int().argmax(dim=1)
+        response_len = supervised.sum(dim=1)
```
[the remark this answers is not in the document]
Comment (kashif) [reply examples/scripts/sft_diffusion_gemma.py]: Good catch. The canvas is now derived from the final assistant span only, matching the reference recipe, and packing raises a clear error.

File: scripts/generate_tiny_models/for_block_diffusion/diffusion_gemma_for_block_diffusion.py
Comment (qgallouedec) [inline scripts/generate_tiny_models/for_block_diffusion/diffusion_gemma_for_block_diffusion.py:1]: Technically, we should have it in a new "for_block_diffusion", not in "for_conditional_generation"

File: examples/scripts/sft_diffusion_gemma.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+        return (loss, outputs) if return_outputs else loss
+
+
+def main(script_args, training_args, model_args):
+    model = DiffusionGemmaForBlockDiffusion.from_pretrained(
+        model_args.model_name_or_path,
+        dtype=model_args.dtype,
+        attn_implementation=model_args.attn_implementation,
+    )
+    tokenizer = AutoTokenizer.from_pretrained(model_args.model_name_or_path)
```
Comment (qgallouedec) [inline examples/scripts/sft_diffusion_gemma.py]: Do we really need to pass the tokenizer? Can't we just rely on the trainer?

**Review by qgallouedec — COMMENTED** [review]
You could add a line about this in sft_trainer.md. Something like "SFTTrainer doesn't natively support diffusion models like gemma-diffusion, but the trainer can be easily extended, see the example"

File: examples/scripts/sft_diffusion_gemma.py [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+        num_blocks = (response_len - 1).clamp(min=0) // block_size + 1
+        block_idx = (torch.rand(batch_size, device=device) * num_blocks).long()
+        encoder_len = prefix_len + block_idx * block_size
+
+        # Clean canvas: the selected block, EOS-filled up to block_size past the end of the response. The fill is
+        # supervised, so every canvas position contributes to the loss.
+        offsets = torch.arange(block_size, device=device)
+        abs_idx = (encoder_len[:, None] + offsets).clamp(max=seq_len - 1)
+        in_response = offsets < (response_len - block_idx * block_size)[:, None]
+        canvas_target = torch.where(in_response, input_ids.gather(1, abs_idx), self.eos_token_id)
```
[the remark this answers is not in the document]
Comment (kashif) [reply examples/scripts/sft_diffusion_gemma.py:232]: Guarded this. Examples clipped to max_length now have their EOS fill dropped from the loss, so we don't teach the model to stop at the truncation boundary, while the surviving response tokens stay supervised.

File: scripts/generate_tiny_models/for_block_diffusion/diffusion_gemma_for_block_diffusion.py
Comment (kashif) [reply scripts/generate_tiny_models/for_block_diffusion/diffusion_gemma_for_block_diffusion.py:1]: Moved it to a new for_block_diffusion dir, thanks.

File: examples/scripts/sft_diffusion_gemma.py (hunk 2 above)
Comment (kashif) [reply examples/scripts/sft_diffusion_gemma.py]: Tried that, but without the tokenizer the trainer auto loads the processor and treats the model as a VLM, which then rejects assistant_only_loss. Passing the tokenizer keeps it on the text path, so I kept it for now.

**Comment by kashif** [comment]
Added a short note in sft_trainer.md pointing to this example, good idea.

File: tests/test_chat_template_utils.py [hunk 4]
```diff
@@ -567,6 +567,9 @@ def test_template_error_returns_false(self):
         pytest.param("trl-internal-testing/tiny-CohereForCausalLM", id="cohere"),
         pytest.param("trl-internal-testing/tiny-Cohere2ForCausalLM", id="cohere2"),
         pytest.param("trl-internal-testing/tiny-DeepseekV3ForCausalLM", id="deepseekv3"),
+        pytest.param(
+            "trl-internal-testing/tiny-DiffusionGemmaForBlockDiffusion", id="diffusion_gemma", marks=require_vision
+        ),
```
Comment (sergiopaniego) [inline tests/test_chat_template_utils.py:595]: something like this for the failing test?

```suggestion
        pytest.param(
            "trl-internal-testing/tiny-DiffusionGemmaForBlockDiffusion", id="diffusion_gemma", marks=[
                  require_vision,
                  pytest.mark.skipif(
                      Version(transformers.__version__) < Version("5.11.0"),
                      reason="DiffusionGemma was introduced in transformers 5.11.0",
                  ),
              ]
        ),
```

**Review by sergiopaniego — APPROVED** [review]
Amazing! lgtm when https://github.com/huggingface/transformers/pull/46572/ is merged

We should add the example to the list of examples in `docs/source/example_overview.md`

Some questions:
- do we have trackio metrics or a checkpoint that we could test?
- what hardware do we need to run it?

File: trl/chat_templates/diffusion_gemma.jinja [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+{%- endfor -%}
+
+{%- if add_generation_prompt -%}
+    {%- if ns.prev_message_type != 'tool_response' and ns.prev_message_type != 'tool_call' -%}
+        {{- '<|turn>model\n' -}}
+        {%- if not enable_thinking | default(false) -%}
+            {{- '<|channel>thought\n<channel|>' -}}
+        {%- endif -%}
+    {%- endif -%}
+{%- endif -%}
```
[the remark this answers is not in the document]
Comment (sergiopaniego) [reply trl/chat_templates/diffusion_gemma.jinja:387]: can you take a look at this @kashif ?

**Review by sergiopaniego — APPROVED** [review]
lgtm after the comment from Cursor's review is reviewed
